language/tools/cost_synthesis/src/module_generator.rs

```diff
@@ -259,23 +259,27 @@ impl ModuleBuilder {
         // Generate a bunch of random function signatures over these types.
         let functions = (0..self.table_size)
             .map(|_| {
-                let num_locals = self.gen.gen_range(1, MAX_NUM_LOCALS);
                 let num_args = self.gen.gen_range(1, MAX_FUNCTION_CALL_SIZE);
+                let num_locals = self.gen.gen_range(num_args, MAX_NUM_LOCALS);
                 let num_return_types = self.gen.gen_range(1, MAX_RETURN_TYPES_LENGTH);
 
-                let locals = (0..num_locals)
+                let args: Vec<SignatureToken> = (0..num_args)
                     .map(|_| {
                         let index = self.gen.gen_range(0, sig_toks.len());
                         sig_toks[index].clone()
                     })
                     .collect();
 
-                let args = (0..num_args)
-                    .map(|_| {
-                        let index = self.gen.gen_range(0, sig_toks.len());
-                        sig_toks[index].clone()
-                    })
-                    .collect();
+                // We make sure that the first `num_args` locals of the function body match the
+                // function signature.
+                let mut locals = args.clone();
+
+                // Now we generate the rest of the locals for the function; these can be of
+                // arbitrary types.
+                for _ in num_args..num_locals {
+                    let index = self.gen.gen_range(0, sig_toks.len());
+                    locals.push(sig_toks[index].clone())
```

Comment:
what? push?

Reply:
The idea being that we have:
Before this loop:

`locals = [arg1Ty, ..., argnTy]`

And, after this loop, that locals will look look like

`local = [arg1Ty, ..., argnTy, randTy1, ...,randTyk]`

This way we can make sure that the function signature and locals signatures match up on their intersection. If that makes sense.

Reply:
Or is the wrong way round? Do I need to be doing a "push_front" on this instead of a push for these later locals signature types?